strawberryfields/api/connection.py

```diff
@@ -301,12 +305,43 @@ def ping(self) -> bool:
             bool: ``True`` if the connection is successful, and ``False`` otherwise
         """
         path = "/healthz"
-        response = requests.get(self._url(path), headers=self._headers)
+        response = self._request("GET", self._url(path), headers=self._headers)
         return response.status_code == 200
 
     def _url(self, path: str) -> str:
         return self._base_url + path
 
+    def _refresh_access_token(self):
+        """ Use the offline token to request a new access token """
+        self._headers.pop("Authorization", None)
+        # TODO: Make sure this is the right path
+        path = "/auth/token"
+        headers = {**self._headers}
+        response = self._request("POST", self._url(path), headers=headers, data={
+            'grant_type': 'refresh_token',
+            'refresh_token': self._token,
+            'client_id': 'public',
+        })
+        if response.status_code == 200:
+            self._headers["Authorization"] = "Bearer {}".format(response.cookies["access_token"])
+        else:
+            raise RequestFailedError(
+                "Authorization failed for request"
+            )
+
+    def _request(self, method: str, path: str, headers: Dict = {}, **kwargs ):
+        """ Wrap all API requests with an auth token refresh if a 401 is recevied
+        from the initial request.
+        """
+        request_headers = {**headers, **self._headers}
+        response = requests.request(method, self._url(path), headers=request_headers, **kwargs)
+        if response.status_code == 401:
```

Comment:
For my own understanding: could we get a 403 here or is it only 401 in every case?

Reply:
No, we should only expect a 401. A 403 would indicate that our token was valid, but the request was rejected because we don't have permission (based on our user profile) to access the requested resource, which isn't something that refreshing the token will fix.